{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e44af920",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-03-25T04:07:08.437802Z",
     "iopub.status.busy": "2023-03-25T04:07:08.437093Z",
     "iopub.status.idle": "2023-03-25T04:07:20.835101Z",
     "shell.execute_reply": "2023-03-25T04:07:20.833641Z"
    },
    "papermill": {
     "duration": 12.405118,
     "end_time": "2023-03-25T04:07:20.837790",
     "exception": false,
     "start_time": "2023-03-25T04:07:08.432672",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "#!pip install datasets accelerate wandb open-clip-torch\n",
    "#!pip install --upgrade diffusers scipy transformers\n",
    "!pip install -qq diffusers transformers datasets accelerate wandb open-clip-torch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ee3dd16c",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-03-25T04:07:20.844528Z",
     "iopub.status.busy": "2023-03-25T04:07:20.843519Z",
     "iopub.status.idle": "2023-03-25T04:07:24.175112Z",
     "shell.execute_reply": "2023-03-25T04:07:24.173856Z"
    },
    "papermill": {
     "duration": 3.33736,
     "end_time": "2023-03-25T04:07:24.177510",
     "exception": false,
     "start_time": "2023-03-25T04:07:20.840150",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[34m\u001b[1mwandb\u001b[0m: W&B API key is configured. Use \u001b[1m`wandb login --relogin`\u001b[0m to force relogin\n",
      "\u001b[34m\u001b[1mwandb\u001b[0m: \u001b[33mWARNING\u001b[0m If you're specifying your api key in code, ensure this code is not shared publicly.\n",
      "\u001b[34m\u001b[1mwandb\u001b[0m: \u001b[33mWARNING\u001b[0m Consider setting the WANDB_API_KEY environment variable, or running `wandb login` from the command line.\n",
      "\u001b[34m\u001b[1mwandb\u001b[0m: Appending key for api.wandb.ai to your netrc file: /root/.netrc\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import wandb\n",
    "wandb_api = '7bc198602eed746dfff8b8b8b95c1a80df1cf705'\n",
    "wandb.login(key=wandb_api)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "3a2bb2a2",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-03-25T04:07:24.183829Z",
     "iopub.status.busy": "2023-03-25T04:07:24.183532Z",
     "iopub.status.idle": "2023-03-25T04:07:35.955126Z",
     "shell.execute_reply": "2023-03-25T04:07:35.953670Z"
    },
    "papermill": {
     "duration": 11.778211,
     "end_time": "2023-03-25T04:07:35.958194",
     "exception": false,
     "start_time": "2023-03-25T04:07:24.179983",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: huggingface_hub in /opt/conda/lib/python3.7/site-packages (0.12.1)\r\n",
      "Collecting huggingface_hub\r\n",
      "  Downloading huggingface_hub-0.13.3-py3-none-any.whl (199 kB)\r\n",
      "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m199.8/199.8 kB\u001b[0m \u001b[31m5.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\r\n",
      "\u001b[?25hRequirement already satisfied: typing-extensions>=3.7.4.3 in /opt/conda/lib/python3.7/site-packages (from huggingface_hub) (4.4.0)\r\n",
      "Requirement already satisfied: packaging>=20.9 in /opt/conda/lib/python3.7/site-packages (from huggingface_hub) (23.0)\r\n",
      "Requirement already satisfied: requests in /opt/conda/lib/python3.7/site-packages (from huggingface_hub) (2.28.2)\r\n",
      "Requirement already satisfied: pyyaml>=5.1 in /opt/conda/lib/python3.7/site-packages (from huggingface_hub) (6.0)\r\n",
      "Requirement already satisfied: tqdm>=4.42.1 in /opt/conda/lib/python3.7/site-packages (from huggingface_hub) (4.64.1)\r\n",
      "Requirement already satisfied: importlib-metadata in /opt/conda/lib/python3.7/site-packages (from huggingface_hub) (4.11.4)\r\n",
      "Requirement already satisfied: filelock in /opt/conda/lib/python3.7/site-packages (from huggingface_hub) (3.9.0)\r\n",
      "Requirement already satisfied: zipp>=0.5 in /opt/conda/lib/python3.7/site-packages (from importlib-metadata->huggingface_hub) (3.11.0)\r\n",
      "Requirement already satisfied: idna<4,>=2.5 in /opt/conda/lib/python3.7/site-packages (from requests->huggingface_hub) (3.4)\r\n",
      "Requirement already satisfied: charset-normalizer<4,>=2 in /opt/conda/lib/python3.7/site-packages (from requests->huggingface_hub) (2.1.1)\r\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /opt/conda/lib/python3.7/site-packages (from requests->huggingface_hub) (2022.12.7)\r\n",
      "Requirement already satisfied: urllib3<1.27,>=1.21.1 in /opt/conda/lib/python3.7/site-packages (from requests->huggingface_hub) (1.26.14)\r\n",
      "Installing collected packages: huggingface_hub\r\n",
      "  Attempting uninstall: huggingface_hub\r\n",
      "    Found existing installation: huggingface-hub 0.12.1\r\n",
      "    Uninstalling huggingface-hub-0.12.1:\r\n",
      "      Successfully uninstalled huggingface-hub-0.12.1\r\n",
      "Successfully installed huggingface_hub-0.13.3\r\n",
      "\u001b[33mWARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv\u001b[0m\u001b[33m\r\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "#!pip install huggingface_hub==0.10\n",
    "!pip install --upgrade huggingface_hub\n",
    "\n",
    "!python -c \"from huggingface_hub.hf_api import HfFolder; HfFolder.save_token('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')\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "87fa37bc",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-03-25T04:07:35.966050Z",
     "iopub.status.busy": "2023-03-25T04:07:35.965670Z",
     "iopub.status.idle": "2023-03-25T04:07:37.204324Z",
     "shell.execute_reply": "2023-03-25T04:07:37.203180Z"
    },
    "papermill": {
     "duration": 1.245373,
     "end_time": "2023-03-25T04:07:37.206814",
     "exception": false,
     "start_time": "2023-03-25T04:07:35.961441",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2023-03-25 04:07:36--  https://github.com/DiningSystem/diffusion-models-class/raw/main/unit2/finetune_model.py\r\n",
      "Resolving github.com (github.com)... 140.82.113.4\r\n",
      "Connecting to github.com (github.com)|140.82.113.4|:443... connected.\r\n",
      "HTTP request sent, awaiting response... 302 Found\r\n",
      "Location: https://raw.githubusercontent.com/DiningSystem/diffusion-models-class/main/unit2/finetune_model.py [following]\r\n",
      "--2023-03-25 04:07:37--  https://raw.githubusercontent.com/DiningSystem/diffusion-models-class/main/unit2/finetune_model.py\r\n",
      "Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...\r\n",
      "Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.\r\n",
      "HTTP request sent, awaiting response... 200 OK\r\n",
      "Length: 5497 (5.4K) [text/plain]\r\n",
      "Saving to: ‘finetune_model.py’\r\n",
      "\r\n",
      "finetune_model.py   100%[===================>]   5.37K  --.-KB/s    in 0s      \r\n",
      "\r\n",
      "2023-03-25 04:07:37 (35.5 MB/s) - ‘finetune_model.py’ saved [5497/5497]\r\n",
      "\r\n"
     ]
    }
   ],
   "source": [
    "!wget https://github.com/DiningSystem/diffusion-models-class/raw/main/unit2/finetune_model.py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "24b2dd04",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-03-25T04:07:37.214982Z",
     "iopub.status.busy": "2023-03-25T04:07:37.214631Z",
     "iopub.status.idle": "2023-03-25T04:07:42.312747Z",
     "shell.execute_reply": "2023-03-25T04:07:42.310958Z"
    },
    "papermill": {
     "duration": 5.105365,
     "end_time": "2023-03-25T04:07:42.315658",
     "exception": false,
     "start_time": "2023-03-25T04:07:37.210293",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cuda\n"
     ]
    }
   ],
   "source": [
    "#export CUDA_VISIBLE_DEVICES=0,1\n",
    "from accelerate import Accelerator\n",
    "accelerator = Accelerator()\n",
    "print(accelerator.device)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "85d99743",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-03-25T04:07:42.326133Z",
     "iopub.status.busy": "2023-03-25T04:07:42.324572Z",
     "iopub.status.idle": "2023-03-25T04:08:31.110377Z",
     "shell.execute_reply": "2023-03-25T04:08:31.109083Z"
    },
    "papermill": {
     "duration": 48.79357,
     "end_time": "2023-03-25T04:08:31.113106",
     "exception": false,
     "start_time": "2023-03-25T04:07:42.319536",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting onnxruntime-gpu\r\n",
      "  Downloading onnxruntime_gpu-1.14.1-cp37-cp37m-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (136.2 MB)\r\n",
      "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m136.2/136.2 MB\u001b[0m \u001b[31m8.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\r\n",
      "\u001b[?25hCollecting coloredlogs\r\n",
      "  Downloading coloredlogs-15.0.1-py2.py3-none-any.whl (46 kB)\r\n",
      "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m46.0/46.0 kB\u001b[0m \u001b[31m3.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\r\n",
      "\u001b[?25hRequirement already satisfied: flatbuffers in /opt/conda/lib/python3.7/site-packages (from onnxruntime-gpu) (23.1.21)\r\n",
      "Requirement already satisfied: packaging in /opt/conda/lib/python3.7/site-packages (from onnxruntime-gpu) (23.0)\r\n",
      "Requirement already satisfied: protobuf in /opt/conda/lib/python3.7/site-packages (from onnxruntime-gpu) (3.20.3)\r\n",
      "Requirement already satisfied: sympy in /opt/conda/lib/python3.7/site-packages (from onnxruntime-gpu) (1.10.1)\r\n",
      "Requirement already satisfied: numpy>=1.21.6 in /opt/conda/lib/python3.7/site-packages (from onnxruntime-gpu) (1.21.6)\r\n",
      "Collecting humanfriendly>=9.1\r\n",
      "  Downloading humanfriendly-10.0-py2.py3-none-any.whl (86 kB)\r\n",
      "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m86.8/86.8 kB\u001b[0m \u001b[31m8.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\r\n",
      "\u001b[?25hRequirement already satisfied: mpmath>=0.19 in /opt/conda/lib/python3.7/site-packages (from sympy->onnxruntime-gpu) (1.2.1)\r\n",
      "Installing collected packages: humanfriendly, coloredlogs, onnxruntime-gpu\r\n",
      "Successfully installed coloredlogs-15.0.1 humanfriendly-10.0 onnxruntime-gpu-1.14.1\r\n",
      "\u001b[33mWARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv\u001b[0m\u001b[33m\r\n",
      "\u001b[0mCollecting onnxruntime.training\r\n",
      "  Downloading onnxruntime_training-1.14.1-cp37-cp37m-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (214.8 MB)\r\n",
      "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m214.8/214.8 MB\u001b[0m \u001b[31m5.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\r\n",
      "\u001b[?25hRequirement already satisfied: h5py in /opt/conda/lib/python3.7/site-packages (from onnxruntime.training) (3.8.0)\r\n",
      "Collecting cerberus\r\n",
      "  Downloading Cerberus-1.3.4.tar.gz (63 kB)\r\n",
      "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m63.8/63.8 kB\u001b[0m \u001b[31m4.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\r\n",
      "\u001b[?25h  Installing build dependencies ... \u001b[?25l-\b \b\\\b \b|\b \b/\b \b-\b \bdone\r\n",
      "\u001b[?25h  Getting requirements to build wheel ... \u001b[?25l-\b \bdone\r\n",
      "\u001b[?25h  Preparing metadata (pyproject.toml) ... \u001b[?25l-\b \bdone\r\n",
      "\u001b[?25hRequirement already satisfied: numpy>=1.16.6 in /opt/conda/lib/python3.7/site-packages (from onnxruntime.training) (1.21.6)\r\n",
      "Requirement already satisfied: flatbuffers in /opt/conda/lib/python3.7/site-packages (from onnxruntime.training) (23.1.21)\r\n",
      "Requirement already satisfied: setuptools>=41.4.0 in /opt/conda/lib/python3.7/site-packages (from onnxruntime.training) (59.8.0)\r\n",
      "Requirement already satisfied: packaging in /opt/conda/lib/python3.7/site-packages (from onnxruntime.training) (23.0)\r\n",
      "Requirement already satisfied: sympy in /opt/conda/lib/python3.7/site-packages (from onnxruntime.training) (1.10.1)\r\n",
      "Requirement already satisfied: onnx in /opt/conda/lib/python3.7/site-packages (from onnxruntime.training) (1.13.1)\r\n",
      "Requirement already satisfied: protobuf in /opt/conda/lib/python3.7/site-packages (from onnxruntime.training) (3.20.3)\r\n",
      "Requirement already satisfied: typing-extensions>=3.6.2.1 in /opt/conda/lib/python3.7/site-packages (from onnx->onnxruntime.training) (4.4.0)\r\n",
      "Requirement already satisfied: mpmath>=0.19 in /opt/conda/lib/python3.7/site-packages (from sympy->onnxruntime.training) (1.2.1)\r\n",
      "Building wheels for collected packages: cerberus\r\n",
      "  Building wheel for cerberus (pyproject.toml) ... \u001b[?25l-\b \bdone\r\n",
      "\u001b[?25h  Created wheel for cerberus: filename=Cerberus-1.3.4-py3-none-any.whl size=58193 sha256=ee5397a7c2ab26535141d2673f070f8d3b024ca069a65008488dd30ac239b300\r\n",
      "  Stored in directory: /root/.cache/pip/wheels/7a/93/69/463dbbac15e3c79384c92ec13f75d46b3e27034e2f764d9f85\r\n",
      "Successfully built cerberus\r\n",
      "Installing collected packages: cerberus, onnxruntime.training\r\n",
      "Successfully installed cerberus-1.3.4 onnxruntime.training-1.14.1\r\n",
      "\u001b[33mWARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv\u001b[0m\u001b[33m\r\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!pip install onnxruntime-gpu\n",
    "!pip install onnxruntime.training"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "327a6456",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-03-25T04:08:31.134473Z",
     "iopub.status.busy": "2023-03-25T04:08:31.134140Z",
     "iopub.status.idle": "2023-03-25T04:08:43.363287Z",
     "shell.execute_reply": "2023-03-25T04:08:43.361680Z"
    },
    "papermill": {
     "duration": 12.242991,
     "end_time": "2023-03-25T04:08:43.366048",
     "exception": false,
     "start_time": "2023-03-25T04:08:31.123057",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  File \"finetune_model.py\", line 121\r\n",
      "    image_pipe.save_pretrained(model_save_name+f'step_{step+1}'push_to_hub=True)\r\n",
      "                                                                         ^\r\n",
      "SyntaxError: invalid syntax\r\n",
      "  File \"finetune_model.py\", line 121\r\n",
      "    image_pipe.save_pretrained(model_save_name+f'step_{step+1}'push_to_hub=True)\r\n",
      "                                                                         ^\r\n",
      "SyntaxError: invalid syntax\r\n",
      "Traceback (most recent call last):\r\n",
      "  File \"/opt/conda/bin/torchrun\", line 8, in <module>\r\n",
      "    sys.exit(main())\r\n",
      "  File \"/opt/conda/lib/python3.7/site-packages/torch/distributed/elastic/multiprocessing/errors/__init__.py\", line 346, in wrapper\r\n",
      "    return f(*args, **kwargs)\r\n",
      "  File \"/opt/conda/lib/python3.7/site-packages/torch/distributed/run.py\", line 762, in main\r\n",
      "    run(args)\r\n",
      "  File \"/opt/conda/lib/python3.7/site-packages/torch/distributed/run.py\", line 756, in run\r\n",
      "    )(*cmd_args)\r\n",
      "  File \"/opt/conda/lib/python3.7/site-packages/torch/distributed/launcher/api.py\", line 132, in __call__\r\n",
      "    return launch_agent(self._config, self._entrypoint, list(args))\r\n",
      "  File \"/opt/conda/lib/python3.7/site-packages/torch/distributed/launcher/api.py\", line 248, in launch_agent\r\n",
      "    failures=result.failures,\r\n",
      "torch.distributed.elastic.multiprocessing.errors.ChildFailedError: \r\n",
      "============================================================\r\n",
      "finetune_model.py FAILED\r\n",
      "------------------------------------------------------------\r\n",
      "Failures:\r\n",
      "[1]:\r\n",
      "  time      : 2023-03-25_04:08:41\r\n",
      "  host      : 0247f4e0ba16\r\n",
      "  rank      : 1 (local_rank: 1)\r\n",
      "  exitcode  : 1 (pid: 182)\r\n",
      "  error_file: <N/A>\r\n",
      "  traceback : To enable traceback see: https://pytorch.org/docs/stable/elastic/errors.html\r\n",
      "------------------------------------------------------------\r\n",
      "Root Cause (first observed failure):\r\n",
      "[0]:\r\n",
      "  time      : 2023-03-25_04:08:41\r\n",
      "  host      : 0247f4e0ba16\r\n",
      "  rank      : 0 (local_rank: 0)\r\n",
      "  exitcode  : 1 (pid: 181)\r\n",
      "  error_file: <N/A>\r\n",
      "  traceback : To enable traceback see: https://pytorch.org/docs/stable/elastic/errors.html\r\n",
      "============================================================\r\n",
      "Traceback (most recent call last):\r\n",
      "  File \"/opt/conda/bin/accelerate\", line 8, in <module>\r\n",
      "    sys.exit(main())\r\n",
      "  File \"/opt/conda/lib/python3.7/site-packages/accelerate/commands/accelerate_cli.py\", line 43, in main\r\n",
      "    args.func(args)\r\n",
      "  File \"/opt/conda/lib/python3.7/site-packages/accelerate/commands/launch.py\", line 831, in launch_command\r\n",
      "    multi_gpu_launcher(args)\r\n",
      "  File \"/opt/conda/lib/python3.7/site-packages/accelerate/commands/launch.py\", line 450, in multi_gpu_launcher\r\n",
      "    raise subprocess.CalledProcessError(returncode=process.returncode, cmd=cmd)\r\n",
      "subprocess.CalledProcessError: Command '['torchrun', '--nproc_per_node', '2', 'finetune_model.py', '--image_size', '256', '--batch_size', '4', '--num_epochs', '100', '--grad_accumulation_steps', '8', '--start_model', 'DiningSystem/xray_diffuser', '--dataset_name', '/kaggle/input/xray2data/test_data_xray2/00001', '--wandb_project', 'dm-finetune', '--log_samples_every', '20', '--save_model_every', '1', '--model_save_name', 'xray_diffuser']' returned non-zero exit status 1.\r\n"
     ]
    }
   ],
   "source": [
    "!accelerate launch --multi_gpu --num_processes=2 finetune_model.py --image_size 256 --batch_size 4 --num_epochs 100\\\n",
    "     --grad_accumulation_steps 8 --start_model \"DiningSystem/xray_diffuser\"\\\n",
    "     --dataset_name \"/kaggle/input/xray2data/test_data_xray2/00001\" --wandb_project 'dm-finetune'\\\n",
    "     --log_samples_every 20 --save_model_every 1 --model_save_name \"xray_diffuser\" "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  },
  "papermill": {
   "default_parameters": {},
   "duration": 108.623039,
   "end_time": "2023-03-25T04:08:44.500167",
   "environment_variables": {},
   "exception": null,
   "input_path": "__notebook__.ipynb",
   "output_path": "__notebook__.ipynb",
   "parameters": {},
   "start_time": "2023-03-25T04:06:55.877128",
   "version": "2.4.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
